src/Game.tsx

```diff
@@ -116,17 +163,80 @@ export function Game({ pokemonOne, pokemonTwo }: Props): JSX.Element {
     };
   }, [isGameStarted]);
 
+  useEffect(() => {
+    const { ballX, ballY } = ballState;
+
+    if (ballX === 0) {
+      if (
+        (ballY >= 50 &&
+          (ballY + MAX_BALL_OFFSET <= player1Coord ||
+            ballY - MIN_BALL_OFFSET >= player1Coord)) ||
+        (ballY <= 50 &&
+          (ballY + MIN_BALL_OFFSET <= player1Coord ||
+            ballY - MAX_BALL_OFFSET >= player1Coord))
+      ) {
+        setGameScore((prevState) => {
+          return {
+            ...prevState,
+            secondPlayerScore: prevState.secondPlayerScore + 1,
+          };
+        });
+        setIsGameStarted(false);
+      }
+    } else if (ballX === 100) {
+      if (
+        (ballY >= 50 &&
+          (ballY + MAX_BALL_OFFSET <= player2Coord ||
+            ballY - MIN_BALL_OFFSET >= player2Coord)) ||
+        (ballY <= 50 &&
+          (ballY + MIN_BALL_OFFSET <= player2Coord ||
+            ballY - MAX_BALL_OFFSET >= player2Coord))
+      ) {
+        setGameScore((prevState) => {
+          return {
+            ...prevState,
+            firstPlayerScore: prevState.firstPlayerScore + 1,
+          };
+        });
+        setIsGameStarted(false);
+      }
+    }
+  }, [ballState, player1Coord, player2Coord]);
+
+  useEffect(() => {
+    if (gameScore.firstPlayerScore === 5 || gameScore.secondPlayerScore === 5) {
+      setBallState(initialBallState);
+      setGameScore(initialScoreState);
+      setPlayer1Coord(initialCardsState);
+      setPlayer2Coord(initialCardsState);
+      setWinner(
+        gameScore.firstPlayerScore > gameScore.secondPlayerScore
+          ? pokemonOne
+          : pokemonTwo
+      );
+      initialBallState.ballStepX = getArrayRandomElement([-2, 2]);
+      initialBallState.ballStepY = getArrayRandomElement([-1, 1]);
```

Comment:
Don't mutate initial State object.
If you wont update `ballStep`, please, use `setState`.